### PR TITLE
empty elementOfJsonPath

### PR DESCRIPTION
**Pull request by vegterb**
return null with empty elementOfJsonPath

### CODE REVIEW

**Comment by fhoeben** [comment]
Looks good. Would you mind adding a test?

**Comment by vegterb** [comment]
JsonFixtureTest Class & twee testen aangemaakt

File: src/main/java/nl/hsac/fitnesse/fixture/slim/JsonFixture.java [hunk 1]
```diff
@@ -73,7 +73,7 @@ public String createFileFromBase64ContentOf(String baseName, String jsonPath) {
 
     public Object elementOfJsonPath(int index, String path) {
         List<Object> all = listJsonPathMatches(path);
-        return all.get(index);
+        return all.size() > 0 ? all.get(index) : null;
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/JsonFixture.java:76]:
```suggestion
        return all.size() > index ? all.get(index) : null;
```

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                    "            \"color\": \"red\",\n" +
+                    "            \"price\": 19.95\n" +
+                    "        }\n" +
+                    "    },\n" +
+                    "    \"expensive\": 10\n" +
+                    "}\n";
+
+    @Test
+    public void testElementOfJsonPathNoResult() {
+        JsonFixture cont = new JsonFixture();
```
Comment (fhoeben) [inline src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:61]: Why not just use `fixture`?

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        cont.load(JSON);
+        Object result = cont.elementOfJsonPath(0,jsonPathNoResults);
+        assertNull(result);
+    }
+
+    @Test
+    public void testElementOfJsonPathResults(){
+        JsonFixture cont = new JsonFixture();
```
Comment (fhoeben) [inline src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:69]: Why not just use `fixture`?

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                    "        }\n" +
+                    "    },\n" +
+                    "    \"expensive\": 10\n" +
+                    "}\n";
+
+    @Test
+    public void testElementOfJsonPathNoResult() {
+        JsonFixture cont = new JsonFixture();
+        cont.load(JSON);
+        Object result = cont.elementOfJsonPath(0,jsonPathNoResults);
```
Comment (fhoeben) [inline src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:63]:
```suggestion
        Object result = cont.elementOfJsonPath(0, jsonPathNoResults);
```

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        assertNull(result);
+    }
+
+    @Test
+    public void testElementOfJsonPathResults(){
+        JsonFixture cont = new JsonFixture();
+        cont.load(JSON);
+        Object result = cont.elementOfJsonPath(0,jsonPathWithResults);
```
Comment (fhoeben) [inline src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:71]:
```suggestion
        Object result = cont.elementOfJsonPath(0, "$.store.book[?(@.category==\"fiction\")].author");
```

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        assertEquals(expectedResult,result);
```
Comment (fhoeben) [inline src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:72]:
```suggestion
        assertEquals(expectedResult, "Evelyn Waugh");
```

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.stream.Stream;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+public class JsonFixtureTest {
+    private JsonFixture fixture = new JsonFixture();
+    private final String jsonPathNoResults = "$.store.book[?(@.category==\"test\")].author";
+    private final String jsonPathWithResults = "$.store.book[?(@.category==\"fiction\")].author";
+    private final String expectedResult = "Evelyn Waugh";
```
Comment (fhoeben) [inline src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:17]: remove

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+public class JsonFixtureTest {
+    private JsonFixture fixture = new JsonFixture();
+    private final String jsonPathNoResults = "$.store.book[?(@.category==\"test\")].author";
+    private final String jsonPathWithResults = "$.store.book[?(@.category==\"fiction\")].author";
```
Comment (fhoeben) [inline src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:16]: remove

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+                    "    },\n" +
+                    "    \"expensive\": 10\n" +
+                    "}\n";
+
+    @Test
+    public void testElementOfJsonPathNoResult() {
+        JsonFixture cont = new JsonFixture();
+        cont.load(JSON);
+        Object result = cont.elementOfJsonPath(0,jsonPathNoResults);
+        assertNull(result);
```
Comment (fhoeben) [inline src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:64]: Add test that getting element 1 also returns `null`

**Review by fhoeben — APPROVED** [review]
Looks good. A few minor points to finish it I believe

File: src/main/java/nl/hsac/fitnesse/fixture/slim/JsonFixture.java (hunk 1 above)
Comment (vegterb) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/JsonFixture.java:76]: Ah, yes... Interesting. This would make sure that when you chose an element which does not exist.

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java (hunk 5 above)
Comment (vegterb) [reply src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:71]: Interesting, why would you keep these values in code and not variables? Does this improve speed? It looked cleaner with the variable, but either way is fine with me.

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java (hunk 6 above)
Comment (vegterb) [reply src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java]: Blindly added, but I expect you wanted to replace the expectedResult variable here.

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java (hunk 2 above)
Comment (vegterb) [reply src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:61]: Agreed

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java (hunk 9 above)
Comment (vegterb) [reply src/test/java/nl/hsac/fitnesse/fixture/slim/JsonFixtureTest.java:64]: Add test with no result, where the jsonPath results in 3 items, but item 3 returns null.

**Comment by vegterb** [comment]
@fhoeben changes made, thanks for the feedback.

**Comment by fhoeben** [comment]
Just remembered `JsonHttpTest`: has its own handling (if only Java had mixins) of getting elements. Please make the same change there. 
Sorry for not mentioning it earlier...

**Comment by vegterb** [comment]
@fhoeben as per your comment I updated `JsonHttpTest` but got a bit stuck on the unit test, my girlfriend suggested Mockito, which was already part of your project so I also updated the JsonFixtureTest. Let me know what you think.

**Comment by fhoeben** [comment]
Although I really like mockito, I try to keep away from using spies as much as possible as they tend to couple your test too tightly to the implementation. Instead I suggest you take a look at wiki/FitNesseRoot/HsacAcceptanceTests/SlimTests/HttpTest/JsonHttpTest.wiki line 30, I believe adding one line there to check element 2 would be the easiest way to test the changed behaviour (line 30 tests the behaviour when the element _is_ present)

**Comment by vegterb** [comment]
I could of course not use the spy implementation and just use a mock for the `HttpResponse` class and keep using all the methodes within the 'JsonHttpTest' class. It's you're call... mock or revert and use:
`|check        |element              |2                   |of json path|airplanes[?(@.airline=='KLM')].type|null|`
is also no problem.

I will revert the changes in the 'JsonFixtureTest'.

**Comment by fhoeben** [comment]
I would prefer `|check |element |2 |of json path|airplanes[?(@.airline=='KLM')].type|null|` 

The spy indeed isolates the specific method, but that method is the implementation detail. The FitNesse test checks the behaviour of the class given a certain response from the remote server, it's completely decoupled from how the fixture arrives at that response. So JsonHttpTest could be completely refactored (for instance by delegating to JsonFixture, or even rewriting it in Kotlin) and the test needs no change: it still verifies whether the correct response is given.

**Comment by vegterb** [comment]
@fhoeben all changes made, should be good now.

File: src/test/java/nl/hsac/fitnesse/fixture/slim/JsonHttpTestTest.java [hunk 10]
```diff
@@ -6,8 +6,7 @@
```
Comment (fhoeben) [inline src/test/java/nl/hsac/fitnesse/fixture/slim/JsonHttpTestTest.java:6]: Please revert all changes to this file (`*` imports are a bad idea, and I like newlines at the end of my files)

**Review by fhoeben — CHANGES_REQUESTED** [review]
Looks good. One minor thing left
